fix(swaps): validate Raiden's resolve request

erkarl opened the pull request:
In this commit we validate Raiden's resolve request to make sure the
payment we're receiving has the previously agreed upon `deal.makerCltvDelta` as the
minimum duration of the lock.

In addition we change the error handling to return a status code 400
(Bad Request) to prevent Raiden from crashing. Previously we returned a
status code of 200 with the response of `{ secret: 'error message' }`.

Closes: https://github.com/ExchangeUnion/xud/issues/1149

kilrau commented:
Please open an issue on raiden side for the crash @erkarl

erkarl commented:
> Please open an issue on raiden side for the crash @erkarl

There already is: https://github.com/ExchangeUnion/raiden/issues/1

kilrau commented:
I successfully tested this branch on native simnet, which since ExchangeUnion/xud#1080 changed the behaviour of the XUD resolver when it gets an unfamiliar rHash and caused raiden to crash. With this branch, this is handled gracefully and raiden doesn't crash.

sangaman commented on lib/raidenclient/types.ts:
```diff
@@ -68,6 +68,7 @@ export type RaidenResolveRequest = {
   reveal_timeout: number;
```
Should we be removing reveal and settle timeout as part of this PR? Or are you planning to do this separately? I'm thinking if it's done together this can be marked as closing #1149.

If so, you can also remove the commented out "unused" fields below.

erkarl replied:
Yes, absolutely - good idea.

sangaman commented on lib/http/HttpServer.ts:
```diff
@@ -67,7 +67,10 @@ class HttpServer {
-      if (err.code === errorCodes.INVALID_ARGUMENT) {
+      if (
+        err.code === errorCodes.INVALID_ARGUMENT ||
+        err.code === swapErrorCodes.INVALID_RESOLVE_REQUEST
```
I'm thinking I should've made this a `switch` statement when I wrote this originally, but now I think it's even more clearly the nicer choice. Would you mind rewriting these if/elses as a switch?

erkarl replied:
Rewrote this as switch - looks much better now, thanks.

sangaman reviewed (commented):
Looks nice, the 400/404/500 HTTP error codes here are much more sensible than returning plain error message strings in place of the preimage. Just a few questions/suggestions below.

sangaman commented on lib/http/HttpServer.ts:
```diff
@@ -67,15 +67,22 @@ class HttpServer {
+          res.writeHead(400);
+          res.end(err.message);
+          break;
+        case swapErrorCodes.INVALID_RESOLVE_REQUEST:
```
This can just flow with the previous case like so we don't need to write the code twice:

```javascript
        case errorCodes.INVALID_ARGUMENT:
        case swapErrorCodes.INVALID_RESOLVE_REQUEST:
          res.writeHead(400);
          res.end(err.message);
          break;
```

sangaman commented on lib/http/HttpServer.ts:
```diff
@@ -67,15 +67,22 @@ class HttpServer {
+          break;
+        default:
+          res.writeHead(500);
+          res.end(JSON.stringify(err));
```
So a `break` here for the last case is not technically necessary, but I kind of like it anyway in case the cases ever get reordered or if we add another case at the end and forget to add the `break`.

There's actually a [tslint rule](https://palantir.github.io/tslint/rules/switch-final-break/) for this I just found that I think might be worth adding, I believe we have the final `break` in use elsewhere in the code so it would only enforce this rule going forward.

sangaman reviewed (changes requested):
Thanks, one last (or maybe two) change re: the new switch statement then this should be good to merge.

sangaman commented on tslint.json:
```diff
@@ -67,6 +67,8 @@
-    "no-switch-case-fall-through": true
+    "no-switch-case-fall-through": true,
+    "switch-final-break": [true, "always"]
+
```
Extra blank line here.

sangaman reviewed (approved):
Looks good! Do you think you could put the tslint rule in a separate commit tho?

Commit:
style: add switch-final-break rule

Commit:
fix(swaps): validate Raiden's resolve request

In this commit we validate Raiden's resolve request to make sure the
payment we're receiving has the previously agreed upon `deal.makerCltvDelta` as the
minimum duration of the lock.

In addition we change the error handling to return a status code 400
(Bad Request) to prevent Raiden from crashing. Previously we returned a
status code of 200 with the response `{ secret: 'error message' }`.

michael1011 commented:
@erkarl @sangaman could you please revisit this PR and merge it asap? The raiden in my local simnet instance is crashing a lot which causes some issues with the balancing of the initial deposit.
